src/ray/raylet/client_connection_test.cc

```diff
@@ -0,0 +1,111 @@
+ public:
+  ClientConnectionTest() : io_service_(), in_(io_service_), out_(io_service_) {
+    boost::asio::local::connect_pair(in_, out_);
+  }
+
+ protected:
+  boost::asio::io_service io_service_;
+  boost::asio::local::stream_protocol::socket in_;
+  boost::asio::local::stream_protocol::socket out_;
+};
+
+TEST_F(ClientConnectionTest, SimpleSyncWrite) {
+  const uint8_t arr[5] = {1, 2, 3, 4, 5};
+  int num_messages = 0;
+
+  ClientHandler<boost::asio::local::stream_protocol> client_handler =
+      [](LocalClientConnection &client) {};
+
+  MessageHandler<boost::asio::local::stream_protocol> message_handler =
+      [&arr, &num_messages](std::shared_ptr<LocalClientConnection> client,
+                            int64_t message_type, const uint8_t *message) {
+        ASSERT_TRUE(!std::memcmp(arr, message, 5));
+        num_messages += 1;
+      };
+
+  auto conn1 = LocalClientConnection::Create(client_handler, message_handler,
+                                             std::move(in_), "conn1");
+
+  auto conn2 = LocalClientConnection::Create(client_handler, message_handler,
+                                             std::move(out_), "conn2");
+
+  RAY_CHECK_OK(conn1->WriteMessage(0, 5, arr));
+  RAY_CHECK_OK(conn2->WriteMessage(0, 5, arr));
+  conn1->ProcessMessages();
+  conn2->ProcessMessages();
+  io_service_.run();
+  ASSERT_EQ(num_messages, 2);
+}
+
+TEST_F(ClientConnectionTest, SimpleAsyncWrite) {
+  const uint8_t msg1[5] = {1, 2, 3, 4, 5};
+  const uint8_t msg2[5] = {4, 4, 4, 4, 4};
+  const uint8_t msg3[5] = {8, 8, 8, 8, 8};
+  int num_messages = 0;
+
+  ClientHandler<boost::asio::local::stream_protocol> client_handler =
+      [](LocalClientConnection &client) {};
+
+  MessageHandler<boost::asio::local::stream_protocol> noop_handler = [](
+      std::shared_ptr<LocalClientConnection> client, int64_t message_type,
+      const uint8_t *message) {};
+
+  std::shared_ptr<LocalClientConnection> reader = NULL;
+
+  MessageHandler<boost::asio::local::stream_protocol> message_handler =
+      [&msg1, &msg2, &msg3, &num_messages, &reader](
+          std::shared_ptr<LocalClientConnection> client, int64_t message_type,
+          const uint8_t *message) {
+        if (num_messages == 0) {
+          ASSERT_TRUE(!std::memcmp(msg1, message, 5));
+        } else if (num_messages == 1) {
+          ASSERT_TRUE(!std::memcmp(msg2, message, 5));
+        } else {
+          ASSERT_TRUE(!std::memcmp(msg3, message, 5));
+        }
+        num_messages += 1;
+        if (num_messages < 3) {
+          reader->ProcessMessages();
+        }
+      };
+
+  auto writer = LocalClientConnection::Create(client_handler, noop_handler,
+                                              std::move(in_), "writer");
+
+  reader = LocalClientConnection::Create(client_handler, message_handler, std::move(out_),
+                                         "reader");
+
+  std::function<void(const ray::Status &)> callback = [this](const ray::Status &status) {
```

Comment:
I added a simple test that checks the status is correctly propagated.

Not sure I follow with the suggestion on passing in writer by value -- this test suite does pass valgrind without leaks so sending messages doesn't leak.

Reply:
Thanks, that test looks good!

What I meant was to use a callback that captures `writer`, so something like this: `auto callback = [this, writer](const ray::Status &status) { ... }`. This is a pretty common pattern in the codebase. Unfortunately, it's also slightly sketchy since we're giving `writer` a `shared_ptr` to itself when we enqueue such a callback, which creates a (temporary) reference cycle. So I was suggesting that we add a simple test to make sure that `writer` will eventually get cleaned up.

Reply:
Oh I see what you mean, added that and it seems fine as well.